src/main/scala/NIC.scala

```diff
@@ -273,27 +278,58 @@ class IceNicRecvPathModule(outer: IceNicRecvPath)
   val io = IO(new Bundle {
     val recv = Flipped(new IceNicRecvIO)
     val in = Flipped(Decoupled(new StreamChannel(NET_IF_WIDTH))) // input stream 
-    val tap = outer.tapFuncs.nonEmpty.option(
-      Vec(outer.tapFuncs.length, Decoupled(new StreamChannel(NET_IF_WIDTH))))
+    val tap = Vec(outer.tapFuncs.length, Decoupled(new StreamChannel(NET_IF_WIDTH)))
     val csum = checksumOffload.option(new Bundle {
       val res = Decoupled(new TCPChecksumOffloadResult)
       val enable = Input(Bool())
     })
-    val buf_free = Output(UInt(8.W))
+    val buf_free = Output(Vec(1 + outer.tapFuncs.length, UInt(8.W)))
   })
 
-  val dropChecks = if (usePauser) Seq(PauseDropCheck(_, _, _)) else Nil
-  val buffer = Module(new NetworkPacketBuffer(
-    inBufFlits, dropChecks = dropChecks, dropless = usePauser))
-  buffer.io.stream.in <> io.in
-  io.buf_free := buffer.io.free
+  def tapOutToDropCheck(tapOut: EthernetHeader => Bool) = {
+    (header: EthernetHeader, ch: StreamChannel, update: Bool) => {
+      val first = RegInit(true.B)
+      val drop = tapOut(header) && first
+      val dropReg = RegInit(false.B)
+
+      when (update && first) { first := false.B; dropReg := drop }
+      when (update && ch.last) { first := true.B; dropReg := false.B }
+
+      drop || dropReg
+    }
+  }
+
+  def duplicateStream(in: DecoupledIO[StreamChannel], outs: Seq[DecoupledIO[StreamChannel]]) = {
+    outs.foreach { out =>
+      out.valid := in.valid
+      out.bits := in.bits
+    }
+    in.ready := outs.head.ready
+    val outReadys = Cat(outs.map(_.ready))
+    assert(outReadys.andR || !outReadys.orR,
+      "Duplicated streams must all be ready simultaneously")
+    outs
+  }
+
+  def invertCheck(check: (EthernetHeader, StreamChannel, Bool) => Bool) =
+    (eth: EthernetHeader, ch: StreamChannel, up: Bool) => !check(eth, ch, up)
+
+  val tapDropChecks = outer.tapFuncs.map(func => tapOutToDropCheck(func))
+  val pauseDropCheck = if (usePauser) Some(PauseDropCheck(_, _, _)) else None
+  val allDropChecks =
+    Seq(tapDropChecks ++ pauseDropCheck.toSeq) ++
+    tapDropChecks.map(check => invertCheck(check) +: pauseDropCheck.toSeq)
+
+  val buffers = allDropChecks.map(dropChecks =>
+    Module(new NetworkPacketBuffer(
+      inBufFlits, dropChecks = dropChecks, dropless = usePauser)))
+  duplicateStream(io.in, buffers.map(_.io.stream.in))
 
-  val tapout = if (outer.tapFuncs.nonEmpty) {
-    val tap = Module(new NetworkTap(outer.tapFuncs))
-    tap.io.inflow <> buffer.io.stream.out
-    io.tap.get <> tap.io.tapout
-    tap.io.passthru
-  } else { buffer.io.stream.out }
+  io.buf_free := buffers.map(_.io.free)
+
+  io.tap <> buffers.tail.map(_.io.stream.out)
```

Comment:
Im slightly confused here, you are assigning the last `tapFuncs.length` buffer outputs to the `io.tap`? Does the order matter here? The `pauseDropCheck` buffer should be the last one right? Shouldn't it be reversed (what is tail should be head and vice versa)?

Reply:
Okay, I guess I should add a comment to this. Each buffer takes a sequence of drop checks and drops the packet if any of them match. So ``allDropChecks`` is a seq of seqs.

The first set of drop checks is "drop if any match the tap function or is a pause frame".

Then the tail has one set of drop checks for each tap. It will drop any packet that doesn't match the tap function or is a pause frame.

Reply:
Ok that makes sense. I missed the `Seq[Seq[]]` part